Add Python interfaces based on nanobind (#796)
LLNL/sundials: doc/superbuild/source/developers/python/index.rst

..
   Author(s): Cody J. Balos @ LLNL
   -----------------------------------------------------------------------------
   SUNDIALS Copyright Start
   Copyright (c) 2002-2025, Lawrence Livermore National Security
   and Southern Methodist University.
   All rights reserved.

   See the top-level LICENSE and NOTICE files for details.

   SPDX-License-Identifier: BSD-3-Clause
   SUNDIALS Copyright End
   -----------------------------------------------------------------------------

.. __Developer.Python:

Python Interfaces
=================

This chapter covers details developers need to know about the SUNDIALS Python interfaces, distributed as the Python package sundials4py.

> We use `nanobind <https://github.com/wjakob/nanobind>__` for the Python bindings. nanobind is a sleeker, faster ``pybind11``.

Reviewer: Is nanobind a fork/version of pybind11?  Is it just a completely separate package?  As written, it sounds like the developers of pybind11 just cleaned things up and changed the name.
```suggestion
We use `nanobind <https://github.com/wjakob/nanobind>__` for the Python bindings. nanobind is a sleeker, faster replacement for ``pybind11``.
```

Author: nanobind is not a fork or version of pybind11. It was written by the primary creator of pybind11 as a sleeker, faster alternative (not a replacement), with the drawback that it is less flexible.

> We use `nanobind <https://github.com/wjakob/nanobind>__` for the Python bindings. nanobind is a sleeker, faster ``pybind11``.

Reviewer 2: Not displaying properly
```suggestion
We use `nanobind <https://github.com/wjakob/nanobind>`__ for the Python bindings. nanobind is a sleeker, faster ``pybind11``.
```

It is a C++ library, i.e. you write your binding code in C++. Nanobind does have some restrictions:

- Cannot bind to functions which take double, or more pointer arguments. I.e., it cannot bind to `**` or `***` and so on. These have to be flattened somehow.
- Cannot implicitly convert between a "View" container class and the underlying C type. I.e., it cannot implicitly convert ``ARKodeView`` to ``void*``.

>    This means user must explicitly convert from the "View" class by calling the ``get`` member function.

Reviewer: ```suggestion
   This means that users must explicitly convert from the "View" class by calling the ``get`` member function.
```

> We use `litgen <https://github.com/pthom/litgen>__` to generate a large portion of the nanobind code.

Reviewer 2: ```suggestion
We use `litgen <https://github.com/pthom/litgen>`__ to generate a large portion of the nanobind code.
```

    **but the binding code generated by the script/litgen falls only under our SUNDIALS license**.
    Because of this, the ``generate.py`` script and litgen extensions are kept in a separate git repository as package,
    ``sundials4py-generator``.

Structure
---------

sundials4py code lives in ``bindings/sundials4py``. The main python module and all of its submodules are defined in ``sundials4py.cpp``.
sundials4py consists of 5 modules. Below we list how each one maps to the directory layout:

- **sundials.arkode**:
  - Implements bindings for all of ARKODE.
  - Source directory: ``arkode/``

- **sundials.cvodes**:
  - Provides bindings for all of CVODES.
  - Source directory: ``cvodes/``

- **sundials.idas**:
  - Contains bindings for all of IDAS.
  - Source directory: ``idas/``

- **sundials.kinsol**:
  - Facilitates bindings for all of KINSOL.
  - Source directory: ``kinsol/``.

- **sundials.core**:
  - All SUNDIALS shared classes/modules and implementations.
  - Source directories:
    - `nvector/`
    - `sunadaptcontroller/`
    - `sunadjointcheckpointscheme/`
    - `sunlinsol/`
    - `sunmatrix/`
    - `sunmemory/`
    - `sunnonlinsol/`
    - `sundials/`

>     - `sundomeigest/`

Reviewer 2: ```suggestion
    - ``nvector/``
    - ``sunadaptcontroller/``
    - ``sunadjointcheckpointscheme/``
    - ``sunlinsol/``
    - ``sunmatrix/``
    - ``sunmemory/``
    - ``sunnonlinsol/``
    - ``sundials/``
    - ``sundomeigest/``
```

Development
-----------

The recommended method for development is to use a typical Python development workflow with ``pip`` rather than invoking CMake directly.

.. code-block:: shell

   cd sundials_root_directory
   python -m venv .venv  # create python virtual environment
   . .venv/bin/activate  # activate the python virtual environment
   pip install scikit-build-core[pyproject] hatchling nanobind # this is a prerequisite for the next step
   MAKEFLAGS="-j$(nproc)" pip install --no-build-isolation -Ceditable.rebuild=true -ve .[dev] # install sundials4py into the virtual environment

> The last ``pip install`` command will allow automatic incremental builds. It will invoke the SUNDIALS `CMake` build system with the

Reviewer 2: ```suggestion
The last ``pip install`` command will allow automatic incremental builds. It will invoke the SUNDIALS ``CMake`` build system with the
```

``-DSUNDIALS_ENABLE_PYTHON=ON`` option through `scikit-build-core <https://scikit-build-core.readthedocs.io/en/latest/index.html>`__.
After the initial build, if you make any changes within SUNDIALS a rebuild will be triggered when you import the ``sundials4py``
module within a Python script.

Different CMake options can be controlled by passing them through the ``--config-settings`` (or ``-C`` for short) option of ``pip install``.
E.g.,

.. code-block:: shell

   MAKEFLAGS="-j$(nproc)" pip install --no-build-isolation -Ceditable.rebuild=true -ve .[dev] \
      -C cmake.define.SUNDIALS_INDEX_SIZE=32

Alternatively, you can set the CMAKE_ARGS environment variable:

.. code-block:: shell

   export CMAKE_ARGS="-DSUNDIALS_INDEX_SIZE=32"
   MAKEFLAGS="-j$(nproc)" pip install --no-build-isolation -Ceditable.rebuild=true -ve .[dev]

Tests
-----

We use pytest for setting up unit/smoke tests of the interfaces. All tests are in ``bindings/sundials4py/test``. The goal is to test the interfacing,
not the correctness of SUNDIALS itself.

--------------

> All user-supplied Python functions have to be wrapped with a functions which converts between a ``std::function`` and a raw C function pointer.

Reviewer: ```suggestion
All user-supplied Python functions have to be wrapped with functions that convert between a ``std::function`` and a raw C function pointer.
```

> This is done by smuggling in a "function table" -- a struct of ``std::function`` members -- in python member of the integrator memory structures, and then storing the integrator memory structure in the ``user_data`` pointer. For the objects which are

Reviewer: ```suggestion
This is done by smuggling in a "function table" -- a struct of ``std::function`` members -- in a ``python`` member inside each integrator memory structure, and then storing the integrator memory structure in the ``user_data`` pointer. For the objects which are
```

module within a Python script.

Different CMake options can be controlled by passing them through the ``--config-settings`` (or ``-C`` for short) option of ``pip install``.
For example, the following builds the interface with 32-bit index types

.. code-block:: shell

   MAKEFLAGS="-j$(nproc)" pip install --no-build-isolation -Ceditable.rebuild=true -ve .[dev] \
      -C cmake.define.SUNDIALS_INDEX_SIZE=32

Alternatively, you can set the CMAKE_ARGS environment variable:

.. code-block:: shell

   export CMAKE_ARGS="-DSUNDIALS_INDEX_SIZE=32"
   MAKEFLAGS="-j$(nproc)" pip install --no-build-isolation -Ceditable.rebuild=true -ve .[dev]

If you are developing new features inside of sundials, you may also need to install the Python code generator package,
`sundials4py-generate <https://github.com/sundials-codes/sundials4py-generate>`__. The steps to install this package are
found in its repo.

Tests
-----

We use pytest for setting up unit/smoke tests of the interfaces. All tests are in ``bindings/sundials4py/test``. The goal is to test the interfacing,
not the correctness of SUNDIALS itself.

> User-supplied functions

Reviewer 3: Add a note about about when/why we need the `StdFn` version of the wrappers (see [this comment](https://github.com/LLNL/sundials/pull/796#discussion_r2603805486)).

Author: This is covered below in the subsection "Special user-supplied functions."

   // the array of N_Vectors and the jcurPtrB output parameter moved to a return
   // value packed into a tuple.
   using CVLsPrecSetupStdFnBS = std::tuple<int, sunbooleantype>(
     sunrealtype t, N_Vector y, std::vector<N_Vector> yS_1d, N_Vector yB,
     N_Vector fyB, sunbooleantype jokB, sunrealtype gammaB, void* user_dataB);

   inline int cvode_lsprecsetupfnBS_wrapper(sunrealtype t, N_Vector y,
                                            N_Vector* yS_1d, N_Vector yB,
                                            N_Vector fyB, sunbooleantype jokB,
                                            sunbooleantype* jcurPtrB,
                                            sunrealtype gammaB, void* user_dataB)
   {
     auto cv_mem   = static_cast<CVodeMem>(user_dataB);
     auto fn_table = get_cvode_fn_table(user_dataB);
     auto fn =
       nb::cast<std::function<CVLsPrecSetupStdFnBS>>(fn_table->lsprecsetupfnBS);
     auto Ns = cv_mem->cv_Ns;

     std::vector<N_Vector> yS(yS_1d, yS_1d + Ns);

     auto result = fn(t, y, yS, yB, fyB, jokB, gammaB, nullptr);

     *jcurPtrB = std::get<1>(result);

     return std::get<0>(result);

>    }

Reviewer 3: It would be good to add a section on things to note when handwriting wrappers. A few things that came up in the review:
* When a wrapper can't be autogenerated and needs to be handwritten (for example [this comment](https://github.com/LLNL/sundials/pull/796#discussion_r2651711550))
* Using `keep_alive` (see [this comment](https://github.com/LLNL/sundials/pull/796#discussion_r2601171788))
* Use `.none()` when we need to allow for passing None (see [this comment](https://github.com/llnl/sundials/pull/796#discussion_r2680217627))

  - Source directory: ``cvodes/``

- **sundials.idas**:
  - Contains bindings for all of IDAS.
  - Source directory: ``idas/``

- **sundials.kinsol**:
  - Facilitates bindings for all of KINSOL.
  - Source directory: ``kinsol/``.

- **sundials.core**:
  - All SUNDIALS shared classes/modules and implementations.
  - Source directories:

    - ``nvector/``
    - ``sunadaptcontroller/``
    - ``sunadjointcheckpointscheme/``
    - ``sunlinsol/``
    - ``sunmatrix/``
    - ``sunmemory/``
    - ``sunnonlinsol/``
    - ``sundials/``
    - ``sundomeigest/``

Development
-----------

sundials4py requires Python 3.12+ and the Interpreter/Development components. E.g., if you were installing Python on a RedHat Linux system,
you could install Python 3.12 with these modules like this:

.. code-block:: shell

   yum install python3.12 python3.12-devel

The recommended method for development is to use a typical Python development workflow with ``pip`` rather than invoking CMake directly.

.. code-block:: shell

   cd sundials_root_directory
   python -m venv .venv  # create python virtual environment
   . .venv/bin/activate  # activate the python virtual environment

>    pip install scikit-build-core[pyproject] hatchling nanobind # this is a prerequisite for the next step

Reviewer 3: To build the interface I had to do
```
pip install scikit-build-core[pyproject] hatchling nanobind setuptools scikit-build pybind11
```

``-DSUNDIALS_ENABLE_PYTHON=ON`` option through `scikit-build-core <https://scikit-build-core.readthedocs.io/en/latest/index.html>`__.
After the initial build, if you make any changes within SUNDIALS a rebuild will be triggered when you import the ``sundials4py``
module within a Python script.

Different CMake options can be controlled by passing them through the ``--config-settings`` (or ``-C`` for short) option of ``pip install``.
E.g.,

.. code-block:: shell

   MAKEFLAGS="-j$(nproc)" pip install --no-build-isolation -Ceditable.rebuild=true -ve .[dev] \
      -C cmake.define.SUNDIALS_INDEX_SIZE=32

Alternatively, you can set the CMAKE_ARGS environment variable:

.. code-block:: shell

   export CMAKE_ARGS="-DSUNDIALS_INDEX_SIZE=32"
   MAKEFLAGS="-j$(nproc)" pip install --no-build-isolation -Ceditable.rebuild=true -ve .[dev]

If you are developing new features inside of sundials, you may also need to install the Python code generator package,
`sundials4py-generate <https://github.com/sundials-codes/sundials4py-generate>`__. The steps to install this package are

> found in its repo.

Reviewer 3: Maybe I'm missing it, but I don't see install instructions in the repo

Author: Added to the README

``nb::arg`` "object". In the example below, a Python user could pass ``None`` for the ``fn`` argument
of ``CVodeRootInit`` since the ``.none()`` call is made.

.. code-block:: cpp

   m.def(
    "CVodeRootInit",
    [](void* cv_mem, int nrtfn, std::function<std::remove_pointer_t<CVRootFn>> fn)
    {
      auto fn_table = get_cvode_fn_table(cv_mem);
      if (fn)
      {
        fn_table->rootfn = nb::cast(fn);
        return CVodeRootInit(cv_mem, nrtfn, &cvode_rootfn_wrapper);
      }
      else { return CVodeRootInit(cv_mem, nrtfn, nullptr); }
    },
    nb::arg("cv_mem"), nb::arg("nrtfn"), nb::arg("fn").none());

Managing Dependent Object Lifetimes
-----------------------------------

Some objects depend on another object existing and being in a valid state. The best
example of this is that all SUNDIALS objects depend on a ``SUNContext`` object. 
We must tell nanobind about these relationships using the ``nb::keep_alive`` function, or in
more complex cases, ``nb::call_policy`` function and ``sundials4py::returns_references_to``
object. Please refer to `https://nanobind.readthedocs.io/en/latest/functions.html#lifetime-annotations`__

> for more information as well as the source file ``bindings/sundials4py/sundials4py_helpers.hpp``.

Reviewer 3: ```suggestion
more complex cases, the ``nb::call_policy`` function and ``sundials4py::returns_references_to``
object. Refer to `the nanobind documentation <https://nanobind.readthedocs.io/en/latest/functions.html#lifetime-annotations>`__
for more information as well as the source file ``bindings/sundials4py/sundials4py_helpers.hpp``.
```
